### ISSUE TITLE
Implement an extendable Context (LEP)

### ISSUE THREAD

**Issue by khorolets** [near/near-lake-framework-rs#54]
### Actuality

Watching for the indexer developers in the wild I’ve noticed some often use-cases we might cover. However, I can’t think of any use case that is required by everyone. Majority? Maybe. Everyone? Definitely not.

I will try to describe the cases and my thoughts on how to handle the problem some might need those features and some might not.

#### Case. Store something in a database
Let’s focus on a couple of things:
- The signature of the user-defined indexing function
- The way we run an indexer

```rust
// The way we run
fn main() -> anyhow::Result<()> {
    // Lake Framework start boilerplate
    near_lake_framework::LakeBuilder::default()
        .mainnet()
        .start_block_height(88444526)
        .build()?
        .run(user_indexer_function) // developer-defined async function that handles each block
}
```

```rust
async fn user_indexer_function(
    mut block: near_lake_primitives::block::Block,
    ctx: near_lake_framework::LakeContext,
) -> anyhow::Result<()> {}
```

Now imagine we have a case where we need to store some data in the database from the user_indexer_function. How to achieve it?

The unpreferred way is to initiate the database connection somewhere globally like lazy_static and then use it in the function.

But we would like to initiate it somewhere in the main and pass it to the user-defined function through the context.

Like this (@frol's suggestion):

```rust
struct MyContext {
    // ...
    lake_context: LakeContext,
}

impl LakeContextExt for MyContext {
    fn new(lake_context: LakeContext) -> Self {
        Self {
            // ...
            lake_context,
        }
    }

    fn lake_context(&mut self) -> &mut LakeContext {
        &mut self.lake_context
    }
}

LakeBuilder::<MyContext>::default()
//           ^^^^^^^^^^^
  .mainnet()
  .build()?
  .run(handle_block_function)

async fn handle_block_function(block: ..., ctx: &mut MyContext) ->anyhow::Result<()> {
  let block_from_rpc = ctx.lake_context().get_block(block.header().hash).await?;
  ...
}
```

And while this might work, there are cases when developers don’t need a context at all, but *they would be forced to provide it even though they don’t need it.*

### The proposal

The main idea of this proposal is to provide the most convenient APIs for both developer groups: the ones who need the context and those who don’t.

#### Contextless API:

```rust
LakeBuilder::default()
  .mainnet()
  .build()?
  .run(user_indexer_function);
  
async fn user_indexer_function(&mut block: Block) -> anyhow::Result<()> {}
```

```rust
struct MyContext {
  pool: Pool<PgSQL>,
}

let my_context = MyContext::new();

LakeBuilder::with_context<MyContext>()
  .mainnet()
  .context(my_context)
  .build()?
  .run(user_indexer_function);
  
async fn user_indexer_function<MyContext>(
  &mut block: Block, 
  ctx: MyContext,
) -> anyhow::Result<()> {}
```

### Action items

- [x] ~~Redesign the `LakeBuilder` to have both `::default()` (no context) and `::with_context<T>()`~~
- [x] Design the `Lake` structure to optionally accept the context `T`
- [x] User-defined indexing function signature should vary based on the presence of the context
- [x] Update existing examples in the repo 
- [x] Create an example or two to show the usage of the context
- [x] Update docs to reflect the changes

**Comment by khorolets** [comment]
cc @pkudinov @telezhnaya @gabehamilton @morgsmccauley @roshaans @eduohe @idea404 @frol

**Comment by khorolets** [comment]
As I am currently in the process of implementing the Enhancement Proposal framework, the present proposal has been published from the internals and assigned the number "001". However, according to the issue number, it should have been designated as "LEP-054". Rest assured that the upcoming proposal will be assigned the appropriate number. I apologize for any inconvenience this may have caused.

**Comment by khorolets** [comment]
Resolved by #68
